[Geng Ning] iP

zognin opened the pull request:
### Duke
> Your friendly task manager

Duke is a task manager. It allows you to

- Add, delete and mark tasks as done
- View your tasks and find them easily

**All you need to do is**

1. Download the JAR release (first release [here](https://github.com/zognin/ip/releases/tag/A-Jar))
2. Double click it
3. Enter commands to add your tasks

If you are a programmer, you can clone the repo to practice Java
Run the main method in `duke.Launcher` to start the program
```java
public class Launcher {
    public static void main(String ...args) {
        Application.launch(Main.class, args);
    }
}
```

**Features** :monocle_face:
- [ ] Manage tasks
- [ ] GUI

Commit:
Add Gradle support

zognin committed:
Implement Level 1 of Duke with Greet, Echo, Exit

zognin committed:
Implement Level 2 Add, List

zognin committed:
Implement Level 3. Mark as Done

zognin committed:
Implement Level 4 ToDos, Events, Deadlines

zognin committed:
Add inputs and expected output to try automated text ui testing and remove unnecessary whitespaces

zognin committed:
Handle errors and use a list object instead of class variable

zognin committed:
Implement Level 6 Delete

zognin committed:
Refactor to separate process into stages and add action type enum

zognin committed:
Improve function that identifies the action type

zognin committed:
Add validation to check if message description is empty before creating action

zognin committed:
Move existing code into packages

zognin committed:
Implement Level-7 Save

zognin committed:
Implement Level 8. Dates and Times

zognin committed:
Merge branch-Level-8 to master

zognin committed:
Refactor to use classes stated in A-MoreOOP

zognin committed:
Organise packages and ui message

zognin committed:
Add JUnit tests for Ui class and some methods in Parser class

zognin committed:
Refactor ui message

zognin committed:
Rename some variables in storage and tasklist

zognin committed:
Add Javadoc to public methods and classes

zognin committed:
Refactor code to meet coding standards in line breaks, packages and naming

zognin committed:
Implement Level 9.Find

zognin committed:
Fix merge conflicts when merging branch-A-JavaDoc and branch-A-CodingStandard

zognin committed:
Fix merge conflicts when merging branch-Level-9 to master

zognin committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

zognin committed:
Change gradle version to 6.3

zognin committed:
Add gradle and checkstyle

zognin committed:
Add javafx dependency

zognin committed:
Implement basic javafx functionality

zognin committed:
Remove ui and simplify messages using vararg

zognin committed:
Add basic greet and exit functionalities

zognin committed:
Add style to dialogbox

zognin committed:
Fix checkstyle

zognin committed:
Use vararg for launcher main

zognin committed:
Update mainClassName in build.gradle

chunweii commented:
LGTM! I can see that you really put in a lot of effort into the iP! Code style is consistent and follows the CS2103/T coding standards. 💯

chunweii commented on src/main/java/controller/DialogBox.java:
```diff
@@ -0,0 +1,68 @@
+import javafx.scene.layout.HBox;
+import javafx.scene.shape.Ellipse;
+
+public class DialogBox extends HBox {
```
Consider adding some JavaDoc here to describe what DialogBox does

chunweii commented on src/main/java/tasklist/Task.java:
```diff
@@ -0,0 +1,127 @@
+     * Supported types of tasks include todo, deadline and event.
+     *
+     * @param description the full user input that may or may not be a valid type of task
+     * @return A new `entity.list.DukeTask`
```
Should `A` be `a` instead?

chunweii commented on src/main/java/tasklist/TaskList.java:
```diff
@@ -0,0 +1,139 @@
+     * Marks task in the list as done, finding the task by its task number.
+     *
+     * @param taskNumber Task number of the task in the list, starting from 1.
+     * @return Task.
```
To be consistent with your other documentations, consider making this into code using markdown ``

chunweii commented on src/main/java/type/DukeCommandTypeEnum.java:
```diff
@@ -0,0 +1,22 @@
+package type;
+
+public enum DukeCommandTypeEnum {
```
Do add a JavaDoc here as well

zognin replied:
Alright I've addressed all your comments, thank you!

JasonC01 reviewed (commented):
Looks really good to me. OOP was well used and I couldn't find any possible flaws. Learnt a lot looking at your code.

zognin committed:
Add storage directory to git ignore

zognin committed:
Add assert to command classes and message

zognin committed:
Rename static variable in storage

zognin committed:
Fix bug to ensure actions are performed according to latest list being viewed

zognin committed:
Refactor commands

zognin committed:
Refactor controllers by adding javadoc and renaming some variables

zognin committed:
Refactor duke by adding javadoc and adjusting spacing

zognin committed:
Refactor exceptions by adding javadoc and renaming ErrorAccessingFileException

zognin committed:
Refactor message and add javadoc for toString method

zognin committed:
Refactor parser and type packages

zognin committed:
Refactor storage and handling of errors in storage

zognin committed:
Refactor tasklist package

zognin committed:
Rename DukeCommandTypeEnum to CommandTypeEnum

zognin committed:
Use streams to format string representation of message

The toString() method combines an array of strings into one string using a for loop with an index to iterate.

It is easier to understand the function by converting the array to a stream then use a for loop. This is because the index is only used for iteration and nothing else.

Let's,
* convert the array of strings to a stream
* iterate through the stream using forEach

zognin committed:
Merge pull request #1 from zognin/branch-A-Assertions

Implement A-Assertions

zognin committed:
Fix merge conflict

zognin committed:
Merge pull request #2 from zognin/branch-A-CodeQuality

Implement A-CodeQuality

zognin committed:
Fix merge conflict

zognin committed:
Merge pull request #3 from zognin/branch-A-Streams

Implement A-Streams in toString() method of Message

zognin committed:
Implement extension C-DetectDuplicates to prevent duplicate tasks from being added

zognin committed:
Add javadoc for enum

zognin committed:
Add gradle.yml for CI

There is no CI pipeline yet. This means a layer of checks before changes to the repo is missing.

A CI pipeline to run gradle check can check for errors in code that passes the repo.

Let's add a gradle.yml file for github actions to run.

zognin committed:
Move gradle.yml

zognin committed:
Merge pull request #4 from zognin/branch-A-CI

Add gradle.yml

bingcheng45 commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,30 @@
+ * Encapsulates main class that starts the application with duke and the user interface.
+ */
+public class Main extends Application {
+    private final Duke duke = new Duke();
```
Maybe for duke can be in all caps to represent a constant.
" Constant names must be all uppercase using underscore to separate words."

zognin replied:
Thanks for point this out!

bingcheng45 commented on src/main/java/tasklist/DeadlineTask.java:
```diff
@@ -0,0 +1,114 @@
+     * @param description Input task string.
+     * @return App representation of a task containing an action description and deadline information.
+     */
+    public static DeadlineTask createTask (String description) throws
```
I think the space between createTask (String ...) can be removed.

bingcheng45 commented on src/main/java/message/Message.java:
```diff
@@ -0,0 +1,46 @@
+     *
+     * @param texts Texts are vararg of strings.
+     */
+    public Message(String ...texts) {
+        assert texts.length > 0 : "A message should have least one string";
+
+        this.texts = texts;
+    }
```
I think it would provide clarity to add in throw AssertionError to indicate there could be potential errors to catch.

zognin replied:
Would it be necessary to throw exceptions for assertions? Because assertions are used to catch programmer mistakes and are separate from exceptions. I would think that if an AssertionError happens, it should be picked up during development, be fixed and not allowed to pass to production? Additionally, they can be disabled?

bingcheng45 reviewed (commented):
LGTM, I love how you utilise what was taught in 2030 concepts such as monads, good use of inheritance, streams etc. Your java conventions are on point as well.

zognin committed:
Add help command

zognin committed:
Rename final variables to all caps

zognin committed:
Improve GUI

zognin committed:
Improve GUI

zognin committed:
Merge branch 'branch-A-BetterGui' into master

zognin committed:
Merge branch 'branch-A-BetterGui' into master

zognin committed:
Handle more storage errors

zognin committed:
Change name of app and size of text and gradle version

zognin committed:
Remove kaomojis as they do not display correctly on some systems

zognin committed:
Refactor commands to return output when execute

zognin committed:
Allow specifying of storage file name

zognin committed:
Add tests for some commands

zognin committed:
Add javadoc for test classes

zognin committed:
Add time for event and extract common time ops

zognin committed:
Change exceptions thrown when creating tasks

zognin committed:
Return more descriptive message if list is empty

zognin committed:
Allow case insensitive search

zognin committed:
Add user guide

zognin committed:
Merge branch 'branch-A-UserGuide' into master

zognin committed:
Add Ui.png

zognin committed:
Change font to helvetica and fix error in message

zognin committed:
Update javadoc

zognin committed:
Fix output messages

zognin committed:
Add validation to disallow start time after end time

zognin committed:
Use a more informative message when there is a missing command description

zognin committed:
Change to use Studio Ghibli released images

zognin committed:
Update incorrect event task storage format message
